// user_service/utils/syncUtils.js
const axios = require('axios');

/**
 * Đồng bộ dữ liệu phụ huynh sang user_service
 */
const syncParentToService = async (parentData) => {
  try {
    await axios.post('http://user_service:5012/api/parents/sync', parentData);
    console.log(`Đồng bộ phụ huynh ${parentData.fullName} sang user_service`);
  } catch (err) {
    console.error('Lỗi đồng bộ phụ huynh sang user_service:', err.message);
  }
};

const deleteParentFromService = async (parentId) => {
  try {
    await axios.delete(`http://user_service:5012/api/parents/sync/${parentId}`);
    console.log(`Đã đồng bộ xóa phụ huynh ${parentId} sang user_service`);
  } catch (err) {
    console.error('Lỗi xóa phụ huynh ở user_service:', err.message);
  }
};

/**
 * Đồng bộ dữ liệu tài xế sang user_service
 */
const syncDriverToService = async (driverData) => {
  try {
    const payload = {
      DriverID: driverData.driverId,
      UserID: driverData.userId,
      Fullname: driverData.fullName,
      PhoneNumber: driverData.phoneNumber,
      Email: driverData.email,
      Status: driverData.status
    };
    await axios.post('http://bus_service:5011/api/drivers', payload);
    console.log(`Đồng bộ tài xế ${driverData.fullName} sang user_service`);
  } catch (err) {
    console.error('Lỗi đồng bộ tài xế sang user_service:', err.message);
  }
};

const deleteDriverFromService = async (driverId) => {
  try {
    await axios.delete(`http://bus_service:5011/api/drivers/${driverId}`);
    console.log(`Đã đồng bộ xóa tài xế ${driverId} sang user_service`);
  } catch (err) {
    console.error('Lỗi xóa tài xế ở user_service:', err.message);
  }
};

/**
 * Đồng bộ dữ liệu phụ huynh sang auth_service
 */
const syncParentToAuth = async (userData) => {
  try {
    await axios.delete(`http://auth_service:5010/api/auth/sync/${userData.userId}`);
    await axios.post('http://auth_service:5010/api/auth/sync', userData);
    console.log(`Đồng bộ User ${userData.userId} sang auth_service`);
  } catch (err) {
    console.error('Lỗi đồng bộ phụ huynh sang auth_service:', err.message);
  }
};

const deleteParentFromAuth = async (userId) => {
  try {
    await axios.delete(`http://auth_service:5010/api/auth/sync/${userId}`);
    console.log(`Đồng bộ xóa User ${userId} khỏi auth_service`);
  } catch (err) {
    console.error('Lỗi xóa user ở auth_service:', err.message);
  }
};

/**
 * Đồng bộ dữ liệu user sang user_service
 */
const syncUserToService = async (userData) => {
  try {
    await axios.post('http://user_service:5012/api/users/sync', userData);
    console.log(`Đồng bộ user ${userData.username} sang user_service`);
  } catch (err) {
    console.error('Lỗi đồng bộ user sang user_service:', err.message);
  }
};

const deleteUserFromService = async (userId) => {
  try {
    await axios.delete(`http://user_service:5012/api/users/sync/${userId}`);
    console.log(`Đã đồng bộ xóa user ${userId} sang user_service`);
  } catch (err) {
    console.error('Lỗi xóa user ở user_service:', err.message);
  }
};

/**
 * Đồng bộ dữ liệu user sang auth_service
 */
const syncUserToAuth = async (userData) => {
  try {
    await axios.delete(`http://auth_service:5010/api/auth/sync/${userData.userID}`);
    await axios.post('http://auth_service:5010/api/auth/sync', userData);
    console.log(`Đồng bộ user ${userData.username} sang auth_service`);
  } catch (err) {
    console.error('Lỗi đồng bộ user sang auth_service:', err.message);
  }
};

const deleteUserFromAuth = async (userId) => {
  try {
    await axios.delete(`http://auth_service:5010/api/auth/sync/${userId}`);
    console.log(`Đồng bộ xóa user ${userId} khỏi auth_service`);
  } catch (err) {
    console.error('Lỗi xóa user ở auth_service:', err.message);
  }
};

module.exports = {
  syncParentToService,
  deleteParentFromService,
  syncDriverToService,
  deleteDriverFromService,
  syncParentToAuth,
  deleteParentFromAuth,
  syncUserToService,
  deleteUserFromService,
  syncUserToAuth,
  deleteUserFromAuth
};
